# Praktikum 4 - Klasifikasi dengan ANN

**Pra Pengolahan Data**

**Langkah 1 - Import Library**

In [18]:
import numpy as np #untuk proses dat anumerik
import pandas as pd #untuk manipulasi data
import tensorflow as tf #untuk pengembangan ANN dan deep learning

**Langkah 2 - Load Data**

In [19]:
dataset = pd.read_csv('Churn_Modelling.csv') # read data
X = dataset.iloc[:, 3:-1].values #ambil dat dari datses untuk fitur dengan memilih kolom yang dimulai dari kolom ke-3 sampai kolom kedua terakhir
y = dataset.iloc[:, -1].values #ambil kolom terakhir sebagai target

In [20]:
print(X)

[[619 'France' 'Female' ... 1 1 101348.88]
 [608 'Spain' 'Female' ... 0 1 112542.58]
 [502 'France' 'Female' ... 1 0 113931.57]
 ...
 [709 'France' 'Female' ... 0 1 42085.58]
 [772 'Germany' 'Male' ... 1 0 92888.52]
 [792 'France' 'Female' ... 1 0 38190.78]]


**Keterangan**
1. Tiap baris mewakili satu baris speerti dalam file CSV
2. Tiap kolom dalam matriks mewakili salah satu fitur dari data. fiur diambil dari kolom ke-3 hingga kedua terakhir
3. Angka dan teks dalam matriks adalah nilai dari atribut 

**Keterangan**



**Langkah 3 - Encoding Data Kategorikal**

In [21]:
from sklearn.preprocessing import LabelEncoder #untuk ubah data labe jadi data numerik 
le = LabelEncoder() #buat objek LabelEncoder yang digunkan untuk melakukan transformasi 
X[:, 2] = le.fit_transform(X[:, 2])#ambil kolom keduda dari fitur X dan mengubahnya nilai dari kolom tsb menjadai bilanagan bulat
print(X)

[[619 'France' 0 ... 1 1 101348.88]
 [608 'Spain' 0 ... 0 1 112542.58]
 [502 'France' 0 ... 1 0 113931.57]
 ...
 [709 'France' 0 ... 0 1 42085.58]
 [772 'Germany' 1 ... 1 0 92888.52]
 [792 'France' 0 ... 1 0 38190.78]]


**Keterangan**

Dari matrik dapat terlihat perbedaan dari hasil matrik seblumnya, nilai matrik saat ini pada kolom 2 nilainya sudah dirubah menjadi bilangan bulat 0 dan 1 yang diconvert dari male dan female.

**Langkah 4 - Encoding Kolom "Geography" dengan One Hot Encoder**

In [22]:
from sklearn.compose import ColumnTransformer #untuk gabungkan beberapa transformasi kolom dalam satu proses 
from sklearn.preprocessing import OneHotEncoder #untuk ubah data kategori jadi representasi biner yang disebut one-hot encoding(jadi vektor biner yang panjangnya sama dengan jumlah kategori yang unik)


#buat objek yang akan menjalankan OneHotEncoder pada kolom ke-1 dari matriks X
#remainder='passthrough' untuk memertahankan kolom lang agar tidak ikut berubah
ct = ColumnTransformer(transformers=[('encoder', OneHotEncoder(), [1])], remainder='passthrough')
X = np.array(ct.fit_transform(X))#use objek untuk ubah matriks X 
print(X)

[[1.0 0.0 0.0 ... 1 1 101348.88]
 [0.0 0.0 1.0 ... 0 1 112542.58]
 [1.0 0.0 0.0 ... 1 0 113931.57]
 ...
 [1.0 0.0 0.0 ... 0 1 42085.58]
 [0.0 1.0 0.0 ... 1 0 92888.52]
 [1.0 0.0 0.0 ... 1 0 38190.78]]


**Keterangan**

Dari matrik tersebut, OneHotEncoder sudah berhasil dijalankan yang ditunjukkan pada kolom bagian negara nerubah menjadi nilai biner 

**Langkah 5 - Split Data**

In [23]:
from sklearn.model_selection import train_test_split #untuk bagi dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0) #bagi data menjadi 4 subset, X adalah data matriks fitur, y adalah array label, 20% dari data digunakan untuk data uji. 

**Langkah 6 - Scaling Fitur**

In [24]:
from sklearn.preprocessing import StandardScaler #untuk penskalaan 
sc = StandardScaler() #buat objek StandarScaler
X_train = sc.fit_transform(X_train) #penskalaan fitur data latih
X_test = sc.transform(X_test) #penskalaan pada data uji

# **Membuat Model ANN**

**Langkah 1 - Inisiasi Model ANN**

In [25]:
ann = tf.keras.models.Sequential()#buat objek model ANN denganpendekatan berurutan 

**Langkah 2 - Membuat Input Layer dan Hidden Layer Pertama**

In [26]:
#menambhakn sebuah Dense ke dalam model ANN 
#lapisan dense -> lapisan penuh terhbung, tiap neuron di lapisan ini terhubung dengan semua neuron di lapisan sebelumnya
#dengan menambah 6 unit neuron 
#relu-> menentukan fungsi aktivasi, relu membantu dalam pemodelan fungsi non linear
ann.add(tf.keras.layers.Dense(units=6, activation='relu')) 

**Langkah 3 - Mmebuat Hidden Layer Kedua**

In [27]:
#menambhakn sebuah Dense ke dalam model ANN 
#lapisan dense -> lapisan penuh terhbung, tiap neuron di lapisan ini terhubung dengan semua neuron di lapisan sebelumnya
#dengan menambah 6 unit neuron 
#relu-> menentukan fungsi aktivasi, relu membantu dalam pemodelan fungsi non linear
ann.add(tf.keras.layers.Dense(units=6, activation='relu'))

**Langkah 4 - Membuat Output Layer**

In [28]:
#menambhakn sebuah Dense ke dalam model ANN 
#lapisan dense -> lapisan penuh terhbung, tiap neuron di lapisan ini terhubung dengan semua neuron di lapisan sebelumnya
#dengan menambah 1 unit neuron 
#sigmoid-> menentukan fungsi aktivasi, sigmoid membantu dalam klasifikasi biner
ann.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

# **Training Model**

**Langkah 1 - Compile Model (Menyatukan Arsitektur) ANN**

In [29]:
# optimizer "adam" untuk mengoptimalkan model Anda selama pelatihan
# 'binary_crossentropy' adalah fungsi kerugian (loss function) untuk ukur kesalahan antara prediksi model dan label yang sebenarnya
# ['accuracy'] adalah metrik evaluasi:mengatur metrik yang ingin dipantau selama pelatihan.
ann.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

**Langkah 2 - Fitting Model**

In [30]:
#Pelatihan ANN denan 100 epoch, pada tiap epoch model akan meminimalkan nilai fungsi kerugian 
ann.fit(X_train, y_train, batch_size = 32, epochs = 100)#batch size-> atur banyak sampel data yang akan digunakan dalam tiap iterasi selama pelatihan 

Epoch 1/100


250/250 [==============================] - 3s 4ms/step - loss: 0.5875 - accuracy: 0.7191
Epoch 2/100
250/250 [==============================] - 1s 4ms/step - loss: 0.4726 - accuracy: 0.7964
Epoch 3/100
250/250 [==============================] - 1s 4ms/step - loss: 0.4442 - accuracy: 0.8009
Epoch 4/100
250/250 [==============================] - 1s 4ms/step - loss: 0.4299 - accuracy: 0.8109
Epoch 5/100
250/250 [==============================] - 1s 4ms/step - loss: 0.4202 - accuracy: 0.8192
Epoch 6/100
250/250 [==============================] - 1s 5ms/step - loss: 0.4107 - accuracy: 0.8263
Epoch 7/100
250/250 [==============================] - 1s 4ms/step - loss: 0.4007 - accuracy: 0.8307
Epoch 8/100
250/250 [==============================] - 1s 3ms/step - loss: 0.3910 - accuracy: 0.8350
Epoch 9/100
250/250 [==============================] - 1s 4ms/step - loss: 0.3820 - accuracy: 0.8416
Epoch 10/100
250/250 [==============================] - 1s 3ms/step - loss: 0.3745 - accuracy: 0.8456
E

**Keterangan**

Dapat dilihat bahwa model berhasil melakukan pelatihan data sebanyak 100 epcoch dan pada epoch ke-100 acuracy menjadi 86% yang nilai akurasi in ilebih tinggi dari saat nilai epoch pertama sebesar 71%. dapat diluhat juga bahwa nilai loss(kerugian) berkurang dri epoch pertama sampi epoch terakhir. 

**Memodelkan Data Baru dan buat Prediksi**

In [31]:
#buat prediksi dengan model ANN yang telah dilatih 
#sc.transform([[1, 0, 0, 600, 1, 40, 3, 60000, 2, 1, 1, 50000]]): merupakan transformasi fitur masukan yang akan diprediksi oleh model. 
#0.5: Ini adalah perbandingan yang digunakan untuk mengubah keluaran model menjadi prediksi biner. Jika hasil prediksi model lebih besar dari 0.5, maka prediksi dianggap sebagai "True" atau 1, sedangkan jika kurang dari atau sama dengan 0.5, maka prediksi dianggap sebagai "False" atau 0.
print(ann.predict(sc.transform([[1, 0, 0, 600, 1, 40, 3, 60000, 2, 1, 1, 50000]])) > 0.5)

1/1 [==============================] - 1s 517ms/step
[[False]]


**Ketrangan**

Hasilnya adalah false yang berarti model memutuskan bahwa data masukan yang diberikan tidak sesuai dengan kategori atau kelas positif yang ditentukan sebelumnya

**Prediksi Dengan Data Testing**

In [32]:
y_pred = ann.predict(X_test)#membuat prediksi pada data uji dengan menggunakna model ANN
y_pred = (y_pred > 0.5) #ubah hasil prediksi menjadi nilai biner. Jika hasil prediksi (probabilitas) lebih besar dari 0.5, maka nilai menjadi True (1), dan jika kurang dari atau sama dengan 0.5, maka nilai menjadi False (0).

#cetak hasil prediksi 
print(np.concatenate((y_pred.reshape(len(y_pred),1), #hasil prediksi yang telah diubah bentuknya menjadi matriks dengan satu kolom
                      y_test.reshape(len(y_test),1)),1)) # label sebenarnya dari data pengujian yang juga diubah bentuknya menjadi matriks dengan satu kolom

 1/63 [..............................] - ETA: 3s

63/63 [==============================] - 0s 3ms/step
[[0 0]
 [0 1]
 [0 0]
 ...
 [0 0]
 [0 0]
 [0 0]]


**Keterangan**

1. Kolom kiri pada matrik merupakan hasil prediksi model. Jika 0 maka model memprediksi false, dan jika 1 berarti model memprediksi true
2. Kolom kanan pada matrik adalah label sebenarnya. Jika 0 berarti data pengujian sebenarnya false, dan 1 jika true. 

Dapt dilihat dari matrik banyak baris dengan pasangan 0 dan 0 yang berarti model meprediksi dengan benar bahwa data tersebut bukan kategori positif. 

**Cek Akurasi dan Confusion Matrix**

In [33]:
from sklearn.metrics import confusion_matrix, accuracy_score #hitung matriks konfusi dan akurasi model 
cm = confusion_matrix(y_test, y_pred)#hitung matriks konfusi 
print(cm)
accuracy_score(y_test, y_pred)#hitung akurasi model 

[[1530   65]
 [ 209  196]]


0.863

**Ketrangan**

1. Terdapat 1530 prediksi yang benar (True Negatives, TN). Merupakan data yang sebenarnya negatif dan telah diprediksi dengan benar sebagai negatif.

2. Terdapat 196 prediksi yang benar (True Positives, TP). Merupakan data yang sebenarnya positif dan telah diprediksi dengan benar sebagai positif.

3. Terdapat 65 prediksi yang salah (False Positives, FP). Merupakan data yang sebenarnya negatif tetapi telah salah diprediksi sebagai positif.

4. Terdapat 209 prediksi yang salah (False Negatives, FN). Merupakan data yang sebenarnya positif tetapi telah salah diprediksi sebagai negatif.

5. Akurasi model adalah 0.863 atau 86.3%. merupakan rasio dari prediksi yang benar (TP + TN) terhadap jumlah total sampel.

KESIMPULAN: 

Secara keseluruhan kinerja model cukup baik yang bisa dilihat dari nilai akurasinya sebesar 86%. 